Storj Backend

calebcase opened the pull request:
#### What is the purpose of this change?

Adding support for Storj as a backend.

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/1571

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on backend/storj/object.go:
```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		return nil, err
+	}
+	defer bucket.Close()
```
You can use `fs.CheckClose` here...

```
func CheckClose(c io.Closer, err *error)
    CheckClose is a utility function used to check the return from Close in a
    defer statement.
```

calebcase commented:
> Hi! Sorry for the delay in responding.
> 
> It looks like you need to vendor your packages - see: https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#adding-a-dependency

Ah, I missed the need for the vendor check in. Fixed

> 
> Rclone supports both modules and non-modules and will do for a while.
> 
> Can you try to fix the lint errors too please?

Yup, fixed

> 
> Did you take a look at https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend - how far did you get down the checklist? Do the integration tests pass?

All the unit tests pass. I'm checking on the integration tests now.

My next step was to start working on the documentation portions.

calebcase commented:
@ncw I also need to sort out the encodings. At the moment I don't do anything with it.

ncw commented:
> > It looks like you need to vendor your packages - see: https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#adding-a-dependency
> 
> Ah, I missed the need for the vendor check in. Fixed

Are you sure that is the minimum requirement?  That included 647 files
which adds 30% in size to the vendored files!  That seems rather
excessive for a single backend!

Compiling rclone with storj adds nearly 7MB to the size of the binary!

Can we use less of storj somehow?  Maybe the API library needs factoring? Maybe we could use rclone's rest module instead of the supplied library?

NB I recently updated all rclone's dependencies so the vendor commit no longer merges.  So you'll need to drop it and rebuild it.

Note that the go.mod and go.sum changes should go in the vendor commit.

> > Rclone supports both modules and non-modules and will do for a while.
> > Can you try to fix the lint errors too please?
> 
> Yup, fixed

Great!

> > Did you take a look at https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend - how far did you get down the checklist? Do the integration tests pass?
> 
> All the unit tests pass. I'm checking on the integration tests now.
> 
> My next step was to start working on the documentation portions.
>
> I also need to sort out the encodings. At the moment I don't do anything with it.

If the backend tests pass then fine.  If not then you'll need to think
about which characters need escaping.

calebcase commented:
> > > It looks like you need to vendor your packages - see: https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#adding-a-dependency
> > 
> > 
> > Ah, I missed the need for the vendor check in. Fixed
> 
> Are you sure that is the minimum requirement? That included 647 files
> which adds 30% in size to the vendored files! That seems rather
> excessive for a single backend!
> 
> Compiling rclone with storj adds nearly 7MB to the size of the binary!

Yup, I've been discussing with the team. There are plans to split out the uplink library into its own repository which should help with the total dependencies and size, but that work hasn't been start afaik.

> 
> Can we use less of storj somehow? Maybe the API library needs factoring? Maybe we could use rclone's rest module instead of the supplied library?

Unfortunately not. Storj is not an http/rest based system (uses gRPC and dRPC transports).

> 
> NB I recently updated all rclone's dependencies so the vendor commit no longer merges. So you'll need to drop it and rebuild it.

:+1: 

> 
> Note that the go.mod and go.sum changes should go in the vendor commit.

I wasn't squashing the commits yet, but I'm happy to flatten them into a logical set if that's desired. Do you want 1 commit total or multiple?

> 
> > > Rclone supports both modules and non-modules and will do for a while.
> > > Can you try to fix the lint errors too please?
> > 
> > 
> > Yup, fixed
> 
> Great!
> 
> > > Did you take a look at https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend - how far did you get down the checklist? Do the integration tests pass?
> > 
> > 
> > All the unit tests pass. I'm checking on the integration tests now.
> > My next step was to start working on the documentation portions.
> > I also need to sort out the encodings. At the moment I don't do anything with it.
> 
> If the backend tests pass then fine. If not then you'll need to think
> about which characters need escaping.

Yup, all the backend tests are passing. I'm sorting through the integration tests (most of which pass). Will update the PR once I've got those remaining ones fixed.

ncw commented:
> > Are you sure that is the minimum requirement? That included 647 files
> > which adds 30% in size to the vendored files! That seems rather
> > excessive for a single backend!
> > Compiling rclone with storj adds nearly 7MB to the size of the binary!
> 
> Yup, I've been discussing with the team. There are plans to split out the uplink library into its own repository which should help with the total dependencies and size, but that work hasn't been start afaik.

Great!  Any idea of timescales for that? If I know it is coming then it isn't a show stopper!

It might be that you can just do a tiny amount of refactoring to chop off most of the dependencies (I don't know because I haven't looked at the dependency graph yet but I've had this experience several times!).

> > NB I recently updated all rclone's dependencies so the vendor commit no longer merges. So you'll need to drop it and rebuild it.
> 
> > Note that the go.mod and go.sum changes should go in the vendor commit.
> 
> I wasn't squashing the commits yet, but I'm happy to flatten them into a logical set if that's desired. Do you want 1 commit total or multiple?

For the final product I'd like two commits
- one adding the backend + docs
- one updating the vendor directory and go.sum/go.mod

I usually like the vendor one first so the project compiles for all commits.

If you fix anything else along the way then put that in a separate commit.

> Yup, all the backend tests are passing. I'm sorting through the integration tests (most of which pass). Will update the PR once I've got those remaining ones fixed.

Fantastic! Ping me when you are done and I'll do some independent testing and give the code fuller review.

I've skimmed through the code - looks good :-)

If you want a debug trace then this is how you do it which will log the entry and the exit of a function

https://github.com/rclone/rclone/blob/e14d968f8d5fdba0d695749648c2b1be05a04fb0/fs/log/log.go#L44

```go
    defer log.Trace(f, "OBJ ls -R ./%s", relative)("err=%v", &err)
```

I note mention of a `storj-sim` network for testing?  Is it easy to test against that?  Rclone has a framework for testing backends against anything which can run in a docker container so if you had a test simulator in a docker container that would be useful!

calebcase commented:
> > > Are you sure that is the minimum requirement? That included 647 files
> > > which adds 30% in size to the vendored files! That seems rather
> > > excessive for a single backend!
> > > Compiling rclone with storj adds nearly 7MB to the size of the binary!
> > 
> > 
> > Yup, I've been discussing with the team. There are plans to split out the uplink library into its own repository which should help with the total dependencies and size, but that work hasn't been start afaik.
> 
> Great! Any idea of timescales for that? If I know it is coming then it isn't a show stopper!

We are actively working on it. It is a bit of a beast, but it necessary due to licensing issues. Until it is complete I don't think we can merge this PR.

> 
> It might be that you can just do a tiny amount of refactoring to chop off most of the dependencies (I don't know because I haven't looked at the dependency graph yet but I've had this experience several times!).

:crossed_fingers: 

> 
> > > NB I recently updated all rclone's dependencies so the vendor commit no longer merges. So you'll need to drop it and rebuild it.
> > 
> > 
> > > Note that the go.mod and go.sum changes should go in the vendor commit.
> > 
> > 
> > I wasn't squashing the commits yet, but I'm happy to flatten them into a logical set if that's desired. Do you want 1 commit total or multiple?
> 
> For the final product I'd like two commits
> 
>     * one adding the backend + docs
> 
>     * one updating the vendor directory and go.sum/go.mod
> 
> 
> I usually like the vendor one first so the project compiles for all commits.
> 
> If you fix anything else along the way then put that in a separate commit.

Sounds good!

> 
> > Yup, all the backend tests are passing. I'm sorting through the integration tests (most of which pass). Will update the PR once I've got those remaining ones fixed.
> 
> Fantastic! Ping me when you are done and I'll do some independent testing and give the code fuller review.
> 
> I've skimmed through the code - looks good :-)

\o/

> 
> If you want a debug trace then this is how you do it which will log the entry and the exit of a function
> 
> https://github.com/rclone/rclone/blob/e14d968f8d5fdba0d695749648c2b1be05a04fb0/fs/log/log.go#L44
> 
> ```go
>     defer log.Trace(f, "OBJ ls -R ./%s", relative)("err=%v", &err)
> ```

That's fantastic!

> 
> I note mention of a `storj-sim` network for testing? Is it easy to test against that? Rclone has a framework for testing backends against anything which can run in a docker container so if you had a test simulator in a docker container that would be useful!

`storj-sim` could be used! It doesn't have a docker container at the moment though. It requires redis and postgresql so it would probably need to be a set of linked services. Other than that there is a dance to set it up and get out credentials for it, but that should be scriptable. If I did get a docker-compose or similar working where would I hook that into the build process?

ncw commented:
> > > Yup, I've been discussing with the team. There are plans to split out the uplink library into its own repository which should help with the total dependencies and size, but that work hasn't been start afaik.
> > 
> > 
> > Great! Any idea of timescales for that? If I know it is coming then it isn't a show stopper!
> 
> We are actively working on it. It is a bit of a beast, but it necessary due to licensing issues. Until it is complete I don't think we can merge this PR.

OK that's fine.  Happy to keep this PR open as long as you need.

> > I note mention of a `storj-sim` network for testing? Is it easy to test against that? Rclone has a framework for testing backends against anything which can run in a docker container so if you had a test simulator in a docker container that would be useful!
> 
> `storj-sim` could be used! It doesn't have a docker container at the moment though. It requires redis and postgresql so it would probably need to be a set of linked services. Other than that there is a dance to set it up and get out credentials for it, but that should be scriptable. If I did get a docker-compose or similar working where would I hook that into the build process?

It is the integration testing that it would need to be hooked in to.

I integration test all the backends daily - results here: https://pub.rclone.org/integration-tests/current/

Storj could be tested by integrating with the live service which is probably what I'd do by default.

However if there was a reasonably easy docker (or other) based test suite then that could be used too.  I haven't quite merged the branch yet, but you can see roughly how it works here: https://github.com/rclone/rclone/tree/testservers/fstest/testserver/init.d - look at one of the Test* files to see what is needed.

ncw commented:
I see a push to this branch :-)  Did you do the slimming down the library?

Thanks

Nick

calebcase commented:
The team is hard at work on splitting out the repos and reducing the external dependencies. I pushed a new update to the branch to sync it with the current upstream, but that doesn't yet have the full repo split. The uplink library itself is undergoing some changes and will mean reworking some of this code to bring it in sync.

ncw commented:
Ping me here when you want me to take a look! I note you'll have rebase on master and and drop your vendor commit(s) and remake them to fix the conflict.

calebcase commented:
@ncw :tada: We are done with the repository rework upstream. This should fix the licensing issues and at many of the dependencies. If you can give the PR review that would be great. Meanwhile we are going to see if there are any more dependencies we can reduce/remove. It looks like we are weighing in at 7MB extra with a plain `go build` and 5MB with `make rclone`.

ncw commented:
> We are done with the repository rework upstream. This should fix the licensing issues and many of the dependencies.

Great

> If you can give the PR a review that would be great.

I've given the code the once over and it looks good :-) Here are my comments

- if you are done with the function entry fs.Debugfs can you remove/comment them
- there doesn't appear to be any docs `docs/content/storj.md`?

How far down the [writing a new backend](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend) checklist did you get? Do the integration tests work OK?

At some point I'll need an account for the integration tester... Are there free accounts, or could you arrange one?

> Meanwhile we are going to see if there are any more dependencies we can reduce/remove. It looks like we are weighing in at 7MB extra with a plain `go build` and 5MB with `make rclone`.

Getting better but still quite a lot :-)

calebcase commented:
> > We are done with the repository rework upstream. This should fix the licensing issues and many of the dependencies.
> 
> Great
> 
> > If you can give the PR a review that would be great.
> 
> I've given the code the once over and it looks good :-) Here are my comments
> 
>     * if you are done with the function entry fs.Debugfs can you remove/comment them

I'd like to leave them for a bit. I still find them very useful for debugging.

> 
>     * there doesn't appear to be any docs `docs/content/storj.md`?

Yup, that's next on my list.

> 
> 
> How far down the [writing a new backend](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend) checklist did you get? Do the integration tests work OK?

All the integration tests are passing. I believe I'm at the `Add your fs to the docs` step.

> 
> At some point I'll need an account for the integration tester... Are there free accounts, or could you arrange one?

I believe we sent you access for an account for the production satellite awhile back. I've also been using storj-sim for local testing, but it requires a postgres. So for a local test environment we need:

* storj-sim
* postgres
* rclone config with a TestStorj entry configured with the output from `storj-sim network env GATEWAY_0_ACCESS` as the access.

> 
> > Meanwhile we are going to see if there are any more dependencies we can reduce/remove. It looks like we are weighing in at 7MB extra with a plain `go build` and 5MB with `make rclone`.
> 
> Getting better but still quite a lot :-)

Yup! Is there a target you'd like us to shoot for?

mniewrzal commented on backend/storj/fs.go:
```diff
@@ -0,0 +1,664 @@
+	if err != nil {
+		fs.Debugf(f, "err: %+v", err)
+
+		return nil, fs.ErrorObjectNotFound
```
it will be not found only if `err == uplink.ErrObjectNotFound`

ncw commented:
I see from the unit tests that various parts of storj.io don't build with go1.12 or go1.13

To fix that you could make them build, or you could use build constraints to stop the backend building for any go below go1.13. (See `backend/sftp` for an example and note that you'll need a file like sftp_unsupported.go to stop the build breaking for unsupported architectures)

Likewise from the compile all architectures test

```
2020/04/21 15:15:16 2 compile failures:
  plan9/386
  plan9/amd64
```

so the build constraint will want to exclude `plan9` also.

So I think that will need build constraints something like

```
// +build go1.13,!plan9
```

And for the negative case (unsupported)

```
// +build !go1.13 plan9
```

I'm going to attempt to have a go with the backend now, get the tests working then answer the remaining questions!

calebcase commented:
> I see from the unit tests that various parts of storj.io don't build with go1.12 or go1.13
> 
> To fix that you could make them build, or you could use build constraints to stop the backend building for any go below go1.13. (See `backend/sftp` for an example and note that you'll need a file like sftp_unsupported.go to stop the build breaking for unsupported architectures)
> 
> Likewise from the compile all architectures test
> 
> ```
> 2020/04/21 15:15:16 2 compile failures:
>   plan9/386
>   plan9/amd64
> ```
> 
> so the build constraint will want to exclude `plan9` also.
> 
> So I think that will need build constraints something like
> 
> ```
> // +build go1.13,!plan9
> ```
> 
> And for the negative case (unsupported)
> 
> ```
> // +build !go1.13 plan9
> ```

Thanks Nick! We were wondering what to do about these. We'll make the adjustments and get back to you.

> 
> I'm going to attempt to have a go with the backend now, get the tests working then answer the remaining questions!

:+1:

ncw commented on backend/tardigrade/object.go:
```diff
@@ -0,0 +1,214 @@
+	for _, option := range options {
+		switch opt := option.(type) {
+		case *fs.RangeOption:
+			s := opt.Start >= 0
```
I think you might find RangeOption.Decode useful here as it returns the offset and length directly

calebcase replied:
Using `RangeOption.Decode` would require knowing the size of the object (and an additional network roundtrip to get it). In theory we could `RangeOption.Decode(0)` under some of the cases, but that implies too much knowledge about the way the decode implementation works.

Maybe we should extend RangeOption with `RequiresSize()` and `SetSize()`? I'm imagining code like:

```
if opt.RequiresSize() {
        object, err := o.fs.project.StatObject(ctx, bucketName, bucketPath)
        if err != nil {
                return nil, err
        }

        opt.SetSize(object.ContentLenth)
}

offset, length := opt.Decode()
```

ncw replied:
Ah, I missed the subtlety that you were avoiding a stat call...

`RequiresSize()` would work. I was thinking of a function to read the size that you could pass in

```
func (o *RangeOption) DecodeSizeFn(size func() (int64, error)) (offset, limit int64, err error) {
}
```

I think `RequiresSize` is probably neater!

I don't think this is something we need to address before merging!

ncw commented on docs/content/overview.md:
```diff
@@ -344,6 +345,7 @@ operations more efficient.
 | QingStor                     | No    | Yes  | No   | No      | Yes     | Yes   | No           | No [#2178](https://github.com/rclone/rclone/issues/2178) | No  | No |
 | SFTP                         | No    | No   | Yes  | Yes     | No      | No    | Yes          | No [#2178](https://github.com/rclone/rclone/issues/2178) | Yes  | Yes |
 | SugarSync                    | Yes   | Yes  | Yes  | Yes     | No      | No    | Yes          | Yes         | No  | Yes |
+| Tardigrade                   | Yes † | No   | No   | No      | No      | Yes   | Yes          | No          | No  | No  |
```
Note that without Move or Copy users won't be able to use the `--backup-dir` functionality of rclone. Almost all the backends support either Move or Copy.

Maybe something for a future improvements to the protocol as I doubt it is easy!

kaloyan-raev replied:
Good point! We plan to implement server-side Copy and Move in Tardigrade, but we are not ready with this yet. This feature is coming in the next months. Until then we will have this limitation.

ncw commented on docs/content/tardigrade.md:
```diff
@@ -0,0 +1,303 @@
+
+Once configured you can then use `rclone` like this.
+
+### Create a new bucket
```
These bits of the docs are a nice quickstart for rclone rather than tardigrade specific...

I wonder whether we should move them into a quickstart page? Or include them in all the backends as a quickstart section?

In an ideal world they'd have links to the full docs...

kaloyan-raev replied:
I wasn't sure how detailed I should be here. We have this info on our Tardigrade doc site: https://documentation.tardigrade.io/how-tos/sync-files-with-rclone

Should I just replace them with the above link?

ncw replied:
If you leave a few basic commands like `ls` and `lsf` and link to the docs that sounds like a good idea. Your docs look very nice :-)

ncw commented:
Let me know in a comment when you want me to take another look :-) Thanks

calebcase commented:
> Normally rclone would retry errors like that with an exponential backoff using the pacer infrastructure, however I notice that you don't appear to be using the pacer - is there a reason for that? I recommend using it as can control the rate of requests and retry things where necessary.

No reason, we just hadn't gotten to that yet. I'll put that next on the list after adding an encoding.

calebcase commented on backend/tardigrade/fs.go:
```diff
@@ -0,0 +1,717 @@
+
+	bn, bp := bucket.Split(path.Join(f.root, relative))
+
+	// NOTE: Technically libuplink does not care about the encoding. It is
```
I tried again to do the encoding, but the problem I think is the test itself: `TestSyncUTFNorm`. Storj can accept any byte sequence as the path. It doesn't care if the input is not in UTF normal form. So, if a user gives rclone non-normal form UTF the user should just get exactly what it sent back. However this test seems to assume that something will do UTF normalization. The test also seems to expect this to happen only on non-Darwin systems. Something feels amiss.

Ideally I'd like to just remove the normalization, have no encoding, and change the test somehow.

ncw replied:
The linux file system works in the same way - it will accept any byte sequence except NUL or `/` in file names so I would expect the test to work here.

This test is really a test of the sync code - it does the unicode normalization and checking itself.

Later on you said you'd like to leave the encoding along - and that is OK with me, so let's leave it like that for the moment.

The rclone standard encoding is pretty minimal - it translates control characters to unicode visible ones and `/` to a unicode equivalent.

Provided your users aren't in the habit of making names with control characters in then this will be unoticeable and is probably a good choice.

calebcase commented on backend/tardigrade/fs.go:
```diff
@@ -0,0 +1,717 @@
+		return nil, err
+	}
+
+	// NOTE: This is necessary because libuplink does not support listings
```
I think the check in `NewFs` is already happening (https://github.com/rclone/rclone/pull/3699/files#diff-a0cba4f5e92efc90c775d7c3915f12cfR246). But maybe we are doing it wrong?

Noted below as well: Removing this causes TestIntegration/FsMkdir/FsPutFiles/FsIsFile to fail.

In theory we could change this to just add the already stat'd object from NewFs, but if the root was deleted or updated since the fs was created it may return stale information. Maybe that's ok though? What do you think?

ncw replied:
I think I see the confusion here..

What is supposed to happen is that in NewFs, if the path points to a file then you return fs.ErrorIsFile but the Fs you return points to the **parent** directory of the file.

That should remove the need for the stat here and fix the `FsIsFile` test hopefully!

calebcase commented:
@ncw We've updated a few things per your feedback and I left a couple comment as well. I have an ever growing wish list of things to add. Do you think what we have here is merge-able (on the assumption that further improvements are on the way)?

Specifically I have the following on my todo list:

* Add handling for the `too many requests` error. Ideally this would happen in libuplink (due to the way retries need to happen at a deeper level inside the piece reconstruction logic of ecclient) or at a minimum at the rclone level with the pacer. If we can get some more time we'd likely do the former.
* Content-Type support (hopefully also using the newly added `--header` stuff from https://github.com/rclone/rclone/issues/59)
* Hash support (and s3:etag as well)
* Support for working directly with the encrypted paths (which is sometimes necessary if different encryption passphrases are used within the same project). It would mirror similar functionality in the uplink cli (`--encrypted`).

The backend seems very usable even without these though.

ncw commented on backend/tardigrade/fs.go:
```diff
@@ -0,0 +1,717 @@
+		return err
+	}
+
+	// NOTE: This is necessary because libuplink does not support listings
```
See above - returning the parent from NewFs along with ErrIsFile should do the trick

calebcase commented:
> > The backend seems very usable even without these though.
> 
> It does :-)

Woot!

> 
> So in summary
> 
>     * we are happy with the default rclone encoding so won't implement our own (I'm assuming you are happy with this)

:+1: yes

>     * we'll address retries, hashes, encrypted paths, headers, Content-Type in future pull requests

Yup

> I think the only bit of code which needs fixing is the NewFs ErrorFileIsFIle handling - hopefully I've understood the problem and put enough hints inline!

Aha! I think I know what needs to happen based on your comments. Thank you for those. I'll do this first thing Monday.

> 
> I think this is looking really good and I'm excited to be merging it :-)

Awesome! We are very excited as well!

calebcase commented:
@ncw

* Fix applied for the NewFs object root.
* PR rebased per two-commit guidance https://github.com/rclone/rclone/pull/3699#issuecomment-556118712
* unit, operation, and sync tests confirmed passing

Should be ready to go!

calebcase commented:
@ncw I check the two test failures, but they don't seem to be related to the new backend?

ncw commented:
> * Fix applied for the NewFs object root.

Looks good to me :-)

> * PR rebased per two-commit guidance [#3699 (comment)](https://github.com/rclone/rclone/pull/3699#issuecomment-556118712)
> * unit, operation, and sync tests confirmed passing

Great!

> Should be ready to go!

Super!

When I ran the tests I got a few failures due to "uplink: too many requests" so I think the pacer is probably the next thing to work on!

> @ncw I check the two test failures, but they don't seem to be related to the new backend?

One of these is #4171 - the other is a new race introduced by 10a6a92e524ac37812913c410218a1c7c1db4446  which I'll fix in a moment!

I'll merge this now :-)

Can I put you down in the [MAINTAINERS file](https://github.com/rclone/rclone/blob/master/MAINTAINERS.md) as the maintainer for the tardigrade backend?

You are welcome to join the rclone organization if you would like - if you would I'll send you an invite. I understand if you've got too much on your plate right now though :-)

ncw commented:
You'll find the new backend in [the latest beta](https://beta.rclone.org) in a few minutes.

I'll release it as part of 1.52 which should be at the weekend :-)

Note that you can find the daily integration tests results here: https://pub.rclone.org/integration-tests/current/

calebcase commented:
> Can I put you down in the MAINTAINERS file as the maintainer for the tardigrade backend?

Totally!

calebcase commented:
> You are welcome to join the rclone organization if you would like - if you would I'll send you an invite. I understand if you've got too much on your plate right now though :-)

I'm happy to join, thanks for the invitation!

ncw commented:
> > Can I put you down in the MAINTAINERS file as the maintainer for the tardigrade backend?
> 
> Totally!

Thank you :-)

> > You are welcome to join the rclone organization if you would like - if you would I'll send you an invite. I understand if you've got too much on your plate right now though :-)
> 
> I'm happy to join, thanks for the invitation!

And done!
